### PR TITLE
8305905: Deprecate String::toLowerCase() and String::toUpperCase()

### PR DESCRIPTION
**Pull request by Glavo**
This PR proposes to deprecate `toLowerCase()`/`toUpperCase()` of `String`.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change requires CSR request [JDK-8305907](https://bugs.openjdk.org/browse/JDK-8305907) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8305905](https://bugs.openjdk.org/browse/JDK-8305905): Deprecate String::toLowerCase() and String::toUpperCase() (**Enhancement** - P4)
 * [JDK-8305907](https://bugs.openjdk.org/browse/JDK-8305907): Deprecate String::toLowerCase() and String::toUpperCase() (**CSR**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13434/head:pull/13434` \
`$ git checkout pull/13434`

Update a local copy of the PR: \
`$ git checkout pull/13434` \
`$ git pull https://git.openjdk.org/jdk.git pull/13434/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13434`

View PR using the GUI difftool: \
`$ git pr show -t 13434`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13434.diff">https://git.openjdk.org/jdk/pull/13434.diff</a>

</details>

### CODE REVIEW

**Comment by Glavo** [comment]
## Motivation

I investigated the usage of these two methods in some Java libraries (including the Java standard library), and **all** of the use cases for these two methods are suspicious.

There are a lot of hidden bugs in these suspicious use cases, which are only exposed in niche locale such as Turkish. For example, [JDK-8301190](https://bugs.openjdk.org/browse/JDK-8301190) is one of the bugs I found.

It can be seen that these two methods are often - or even always - misused, resulting in hidden bugs. I think it is necessary to address these two sources of trouble


## Possible Solutions

This PR is currently a draft and requires further discussion to determine the final solution.

Here are several solutions:

### Do nothing

We have detailed the dangers of these two methods in the Javadoc, so users should take responsibility for themselves.

But this does not solve any problems, there will still be a large number of users who misuse them.

### Change their behavior to make them locale insensitive

The advantage of this solution is that there are already many use cases that abuse them, so directly change their behavior can solve these hidden bugs without the user modifying the code.

But it's a breaking change, breaking code that was using them correctly before, even though there might be few of them.

Also, if this is done, users of higher versions of the JDK will not get warnings, even though they may want their code to run on lower versions of JRE.

### Just deprecate them

We can require users to always use `toLowerCase(Locale)`/`toUpperCase(Locale)`.

Although this is feasible, users may get tired of the verbose writing of `toLowerCase(Locale.ROOT)`.

### Deprecate them, and provide locale insensitive alternative

This is the current implementation of this draft.

This draft currently provides two new APIs: `lowercase()` and `uppercase()`, which are aliases for `toLowerCase(Locale.ROOT)` and `toUpperCase(Locale.ROOT)` respectively.

We recommend users to use these two new APIs as they are locale insensitive. If users want locale-sensitive case conversion, they need to call `toLowerCase(Locale)`/`toUpperCase(Locale)` explicitly.


These are only drafts at the moment, I'd love your feedback to improve it, thanks!

File: src/java.base/share/classes/java/lang/String.java [hunk 1]
```diff
@@ -3618,15 +3641,38 @@ public String toUpperCase(Locale locale) {
...
+     * Converts all of the characters in this {@code String} to upper
+     * case using the rules of the invariant locale. This method is equivalent
+     * to {@code toUpperCase(Locale.ROOT)}.
+     *
+     * @return  the {@code String}, converted to uppercase.
+     * @see     java.lang.String#toUpperCase(Locale)
+     * @since   21
+     */
+    public String uppercase() {
+        return toUpperCase(Locale.getDefault());
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/lang/String.java:3673]:
```suggestion
        return toUpperCase(Locale.ROOT);
```

**Comment by liach** [comment]
/csr

1. You can fix the trailing whitespace in intellij idea by formatting the problematic code
2. Many specifications link to your deprecated `toUpperCase` and `toLowerCase` methods, like those in `Character`. They need to be updated as well.
3. A peek into `Locale.ROOT` reveals it was added only in JDK 6. Might mention that into the deprecation reason as well.

**Comment by merykitty** [comment]
As `Locale.ROOT` exists already, this ticket should only concern deprecating `toLowerCase()` and `toUpperCase()` and leave adding locale independent alternatives for a separate issue. Also please add a CSR and issue the CSR command here. Thanks.

**Comment by eirbjo** [comment]
> As `Locale.ROOT` exists already, this ticket should only concern deprecating `toLowerCase()` and `toUpperCase()` and leave adding locale independent alternatives for a separate issue.

+1 These are two separate issues and should be handled independently.

**Comment by Glavo** [comment]
I have updated the PR and only retained the deprecation of these two methods.

After this PR is merged, I will create a new API in a new PR and gradually replace the old usage in the standard library.

**Comment by eirbjo** [comment]
Even though your commits will be squashed into one before integration, I think it would be good if you use more descriptive commit messages than just "commit". This will make it easier for reviewers to follow and review your progress in this PR.

**Comment by eirbjo** [comment]
If you want this PR to be a draft, you could use the "Convert to draft" link to make it a proper draft PR.

**Comment by merykitty** [comment]
Here is your ticket: https://bugs.openjdk.org/projects/JDK/issues/JDK-8305905
And the CSR: https://bugs.openjdk.org/projects/JDK/issues/JDK-8305907

Please fix the failures observed in GHA.

**Comment by liach** [comment]
> > 1. You need to check all use cases of `toLowerCase()` and `toUpperCase()` in the whole JDK (excluding tests); either convert them to `toLowerCase(Locale.ROOT)` or add `@SuppressWarnings("deprecation")`.
> 
> I have a slight preference for simply adding `@SuppressWarnings("deprecation")` in this PR. That way this PR can focus on the deprecation issue only, and we will not need to consider risks of behavioral changes.

We can actually just explicitly switch to `toLowerCase(Locale.getDefault())` when the purpose is clear; but indeed there will be some confusing cases that are better handled by engineers in each area.

**Comment by Glavo** [comment]
> I have a slight preference for simply adding `@SuppressWarnings("deprecation")` in this PR. That way this PR can focus on the deprecation issue only, and we will not need to consider risks of behavioral changes.

This requires modifying many source files in many modules. I'm worried that others may have objections to doing this, which has led me to do a lot of things in vain. So I thought maybe it would be better to wait for the CSR to be approved before doing it?

**Comment by liach** [comment]
> > I have a slight preference for simply adding `@SuppressWarnings("deprecation")` in this PR. That way this PR can focus on the deprecation issue only, and we will not need to consider risks of behavioral changes.
> 
> This requires modifying many source files in many modules. I'm worried that others may have objections to doing this, which has led me to do a lot of things in vain. So I thought maybe it would be better to wait for the CSR to be approved before doing it?

To have the CSR approved, you need to fill it in first. If you provide the content, I can help you fill it in. Then you need an engineer (a reviewer I think) review before finalizing it for review by Joe Darcy. 

I recommend you just do the changes (add SuppressWarnings) now, as:
1. build depends on it
2. they are implementation detail that don't go through csr. Look at the deprecation of SecurityManager example.

**Comment by Glavo** [comment]
> To have the CSR approved, you need to fill it in first. If you provide the content, I can help you fill it in.

Thank you very much! Here is the content:

---

## Summary

Deprecate `String::toLowerCase()` and `String::toUpperCase()`, users should explicitly pass the `Locale`.

## Problem

The `String::toLowerCase()` and `String::toUpperCase()` without parameters are locale sensitive, and they are widely misused in places that should be local insensitive.

When they are misused, they may produce unexpected results.

For instance, `"TITLE".toLowerCase()` in a Turkish locale returns `"t\u0131tle"` where `'\u0131'` is the LATIN SMALL LETTER DOTLESS I character.

## Solution

Deprecate the two methods and recommend using the corresponding `String::toLowerCase(Locale)`, `String::toUpperCase(Locale)`.

**Comment by eirbjo** [comment]
> Thank you very much! Here is the content:

It seems your CSR is missing the Specification section. This could be just an inline patch file for the String.java API changes.

**Comment by Glavo** [comment]
> It seems your CSR is missing the Specification section. This could be just an inline patch file for the String.java API changes.

merykitty already filled out this section for me when creating the CSR, so there are only the missing part.

**Comment by eirbjo** [comment]
> That's fine, although a patch formatted specification section would make it even easier for CSR reviewers to see what actually changed.

Nevermind, the Specification section is already pretty much in a patch format anyway :-)

**Comment by eirbjo** [comment]
```
 * <p>
 * Instead, to obtain correct results for locale-insensitive strings, use
 * {@code toUpperCase(Locale.ROOT)}; otherwise, use
 * {@code toUpperCase(Locale.getDefault())}.
 ```
This last sentence of the `@deprecated` notes seems punctuation-heavy. Would it be possible to drop the leading `Instead, ` and just go with `To obtain correct..`

**Comment by eirbjo** [comment]
I think the `@deprected` note needs a bit of work:

> This method is locale sensitive

What does "locale sensitive" mean? Sensitive to which locale?

> and may produce unexpected results if used for strings that are intended to be interpreted locale independently.

Independently of which locale?

> Instead, to obtain correct results for locale insensitive strings, use `{@code toLowerCase(Locale.ROOT)}`;

What is a "locale insensitive string"? Insensitive to which locale?
 
> otherwise, use {@code toLowerCase(Locale.getDefault())}.

The "otherwise" case here makes it seem this is for people who want "incorrect results", which is probably nobody :-)

I think the text needs to be more specific about which locale it is referring to, both when referring to the "default locale"  and the "case insensitive locale", which I guess is the "language/country neutral locale" obtained by `Locale.ROOT`

Here is an attempt to be more specific, I'm not claiming this is perfect, and I'm not a native English speaker:

```
     * This method is sensitive to the default locale for this instance
     * of the Java Virtual Machine. It  may produce unexpected
     * results when used on strings that are intended to be processed 
     * independent of the default locale.
     * 
     * <p>
     * To obtain language/country neutral results, use
     * {@code toLowerCase(Locale.ROOT)}; otherwise, use
     * {@code toLowerCase(Locale.getDefault())}.
```

**Comment by eirbjo** [comment]
> I think the `@deprected` note needs a bit of work:

I see now that most of the `@deprected` clause is lifted from the API description of these methods.

Maybe it would be good to get this PR and the CSR in a state where it is ready for review, then more reviewers can give their inputs on the text.

**Comment by Glavo** [comment]
> Maybe it would be good to get this PR and the CSR in a state where it is ready for review, then more reviewers can give their inputs on the text.

I am adding `@SuppressWarnings("deprecation")` everywhere, but now if I push them, key changes will be drowned out by minor changes in dozens of files. I'm hesitating if I should have more people review the changes to String first.

> Here is an attempt to be more specific, I'm not claiming this is perfect, and I'm not a native English speaker:

I think I also need more opinions on the deprecation section in Javadoc. 

I am not a native English speaker either, so it is important for me to have more opinions from those who are good in English.

**Comment by eirbjo** [comment]
> I am adding `@SuppressWarnings("deprecation")` everywhere, but now if I push them, key changes will be drowned out by minor changes in dozens of files. I'm hesitating if I should have more people review the changes to String first.

If this is a real worry, then I would consider doing as Roger mentioned in his recent response on `core-libs-dev`:

> The most positive suggestion from the January thread [1] was to fix the uses in the JDK in small batches and carefully review each to make sure there are no unintended consequences

Then perhaps consider adding new methods, and finally consider the deprecation issue.

Perhaps it is better to fix the JDK first, then offer better alternatives, then finally consider deprecation as the last step.

**Comment by Glavo** [comment]
> Perhaps it is better to fix the JDK first, then offer better alternatives, then finally consider deprecation as the last step.

This is worth considering.

I will start these tasks as soon as possible. I hope to get them in Java 21, so it's best to complete them before entering Rampdown Phase.

**Pull request closed without merging**